[doc] Logstash Kubernetes - Securing Logstash (#14737)
elastic/logstash: docsk8s/setting-up/ls-k8s-secure.asciidoc

In order to prepare your environment to be production ready, you'll need to set up secure communication between each of your Elastic resources.

[[security-tls]]
==== Setting up TLS

> It's important to set up transport layer security (TLS) to ensure safe communication between the Elastic Stack components running in {k8s}.

Reviewer: ```suggestion
Transport layer security (TLS) helps ensure safe communication between the {stack} components running in {k8s}.
```

> Let's take {filebeat} and {ls} TLS mutual verification as an example. {ls} serves as the server side, while {filebeat} is the client.

Reviewer 2: I wonder if we should either link to the [securing communication between beats and logstash page](https://www.elastic.co/guide/en/beats/filebeat/current/configuring-ssl-logstash.html) or have the same caveat that "Creating a correct SSL/TLS infrastructure is outside the scope of this document" that that document does

Author: 👍 link to filebeat doc

> Create a `Secret` containing server and client SSH keys:

Reviewer 2: Typo

```suggestion
Create a `Secret` containing server and client SSL keys:
```

[source,sh]
--
kubectl create secret generic logstash-beats-tls --from-file=ca.crt --from-file=client.crt --from-file=client.key --from-file=server.crt --from-file=server.pkcs8.key
--

> On {ls}, configurate the server certificates to the pipeline:

Reviewer 2: ```suggestion
On {ls}, configure the server certificates to the pipeline:
```

      volumes:
        - name: logstash-beats-tls
          secret:
            secretName: logstash-beats-tls
--

On {filebeat}, configurate the client certificates:

[source,yaml]
--
apiVersion: beat.k8s.elastic.co/v1beta1
kind: Beat
metadata:
  name: demo
spec:
  type: filebeat
  config:
    output.logstash:
      ssl.certificate_authorities: ["/usr/share/filebeat/ca.crt"]
      ssl.certificate: "/usr/share/filebeat/client.crt"
      ssl.key: "/usr/share/filebeat/client.key"
(...)
  deployment:
    podTemplate:
      spec:
        containers:
          - name: filebeat
            volumeMounts:
              - name: logstash-beats-tls
                mountPath: /usr/share/filebeat/ca.crt
                subPath: ca.crt
              - name: logstash-beats-tls
                mountPath: /usr/share/filebeat/client.key
                subPath: client.key
              - name: logstash-beats-tls
                mountPath: /usr/share/filebeat/client.crt
                subPath: client.crt
        volumes:
          - name: logstash-beats-tls
            secret:
              secretName: logstash-beats-tls
--

[[security-k8s-secret]]
==== Using secret

> One way to config confidential data, such as password, is to use `Secret` and references it as container environment variable.

Reviewer 2: ```suggestion
To store sensitive information, such as a password, we can use a {k8s} `Secret`, and reference it as a container environment variable.
```

> 

Reviewer 2: We should probably add a note here that this is for illustration purposes and, in production, managing k8s secrets should be done using recognized [good practices](https://kubernetes.io/docs/concepts/security/secrets-good-practices/) to ensure the protection of sensitive information.

> Encode confidential data with Base64:

Reviewer 2: Add note that `Base64` is encoding, not encryption?

              valueFrom:
                secretKeyRef:
                  name: logstash-secret
                  key: ES_PW
--

[[security-logstash-keystore]]
==== Using the {ls} keystore

{ls} can use the key of {logstash-ref}/keystore.html[keystore] in place of the confidential data when configure sensitive settings.

To create `Secret` from an existing keystore `logstash.keystore`:

[source,sh]
--
kubectl create secret generic logstash-keystore --from-file=logstash.keystore --dry-run=client -o yaml
--

Mount the `Secret` to the {ls} config directory in `Deployment`:

[source,yaml]
--
apiVersion: apps/v1
kind: Deployment
(...)
    spec:
      containers:
        - name: logstash
          env:
            - name: LOGSTASH_KEYSTORE_PASS <1>
              valueFrom:
                secretKeyRef:
                  name: logstash-secret
                  key: LOGSTASH_KEYSTORE_PASS
(...)
          volumeMounts:
            - name: logstash-keystore
              mountPath: /usr/share/logstash/config/logstash.keystore
              subPath: logstash.keystore
      volumes:
        - name: logstash-keystore
          secret:
            secretName: logstash-keystore
--

> <1> `LOGSTASH_KEYSTORE_PASS` is required only when the keystore is protected by {logstash-ref}/keystore.html#keystore-password[password]

Reviewer 2: Maybe remove "only", as we recommend the use of password protection in production

```suggestion
<1> `LOGSTASH_KEYSTORE_PASS` is required when the keystore is protected by {logstash-ref}/keystore.html#keystore-password[password]
```

[[security-eck-secrets]]
==== Securing connection to {es} on ECK

[[security-eck-secrets-pw]]
===== Authentication

> ECK create a user for each Elastic resources. To access the resources, such as {es}, {ls} needs username and password. 

Reviewer 2: ```suggestion
ECK creates a user for every Elastic resource. To access these resources, such as {es}, {ls} needs a username and password. 
```
